materialize-boilerplate/apply.go

```diff
@@ -0,0 +1,229 @@
+	NewProjections []pf.Projection
+
+	// NewlyNullableFields is the fields in the endpoint that should now be made nullable, that
+	// weren't before. They may need to be nullable because a materialized field is now nullable and
+	// wasn't before, or because they exist as non-nullable in the destination and aren't included
+	// in the materialization's field selection.
+	NewlyNullableFields []EndpointField
+
+	// NewlyDeltaUpdates is if the materialized was standard updates per the previously applied
+	// materialization spec, and is now not. Some systems may need to do things like drop primary
+	// key restraints in response to this change.
+	NewlyDeltaUpdates bool
+}
+
+// Applier represents the capabilities needed for an endpoint to apply changes to materialized
+// resources based on binding changes. Many of these functions should return an ActionApplyFn, which
+// will be executed only if the `Apply` is not a dry-run.
+type Applier interface {
+	// CreateMetaTables is called to create the tables (or the equivalent endpoint concept) that
+	// store a persisted spec and any other metadata the materialization needs to persist.
+	CreateMetaTables(ctx context.Context, spec *pf.MaterializationSpec) (string, ActionApplyFn, error)
+
+	// LoadSpec loads the persisted spec from the metadata table.
+	LoadSpec(ctx context.Context, materialization pf.Materialization) (*pf.MaterializationSpec, error)
+
+	// PutSpec upserts a spec into the metadata table.
+	PutSpec(ctx context.Context, spec *pf.MaterializationSpec, version string, exists bool) (string, ActionApplyFn, error)
+
+	// CreateResource creates a new resource in the endpoint. It is called only if the resource does
+	// not already exist.
+	CreateResource(ctx context.Context, spec *pf.MaterializationSpec, bindingIndex int) (string, ActionApplyFn, error)
+
+	// ReplaceResource replaces a resource in the endpoint so that it can start materializing from
+	// the beginning. It may be called even if the resource doesn't already exist. This is usually
+	// carried out with a "CREATE OR REPLACE ..." type of action.
+	ReplaceResource(ctx context.Context, spec *pf.MaterializationSpec, bindingIndex int) (string, ActionApplyFn, error)
+
+	// UpdateResource updates an existing resource. The `BindingUpdate` contains specific
+	// information about what is changing for the resource. `NewProjections` are assured to not
+	// already exist in the destination, and `NewlyNullableFields` are assured to non-nullable in
+	// the destination. It's called for every binding, although it may not have any `BindingUpdate`
+	// parameters. This is to allow materializations to perform additional specific actions on
+	// binding changes that are not covered by the general cases of the `BindingUpdate` parameters.
+	UpdateResource(ctx context.Context, spec *pf.MaterializationSpec, bindingIndex int, applyParams BindingUpdate) (string, ActionApplyFn, error)
```

Comment:
nit: applyParams => bindingUpdate ?

Should it be pluralized, along with the struct itself (there can be _many_ updates for the binding)?

Reply:
I'll change `applyParams` to `bindingUpdate` since `applyParams` was a holdover from the prior SQL-only version of this. `bindingUpdate` as singular is still making more sense to me though.